[BD-03] [TNL-7350] [BB-2524] Discussion configuration API

xitij2000 opened the pull request:
This PR adds a new abstract model to store configuration options in `CourseAppConfigOptionsModel`. This new model is designed to make configuration options defined such that configuration options can be defined at different levels and are accordingly available in different contexts. The aim is to allow different sets of people to design the settings vs those who use them.

For instance, a configuration defined for the site will be available globally on that site. A configuration defined for an org can only be available for courses (or potentially other content) in that org. Likewise, configurations can be restricted to a particular course or course run. This can allow, for instance, an admin to pre-populate LTI or other secret configuration, which can then be selected by the relevant course authors from a list without needing access to the secret configuration. 

Each configuration option has a name and description which is meant to be displayed on a UI, and a slug that is designed for internal reference. Configurations are always referenced via their slug, which makes it possible to update the settings without needing to change the reference in multiple places, there is a mechanism to fetch the current/latest version of the config. 

This PR also utilises this new abstract model to create a new discussion configuration model, and APIs to access them. The discussion configuration model adds on a `provider` field and two config fields `config`, and `private_config`. The provider field is meant to reference the discussion plugin that this configuration is referencing. The actual configuration is divided into config and private_config so any secrets such as API keys, client ids, and client secrets etc can be put in private config while the regular config object can contain other configurable parameters for the config. 

**JIRA tickets**: https://openedx.atlassian.net/browse/TNL-7350

**Dependencies**: None

**Sandbox URL**: TBD - sandbox is being provisioned.

**Merge deadline**: "None"

**Testing instructions**:

There is no UI or external API at the moment, so automates tests should do it.

**Author notes and concerns**:

The final location of this code is still uncertain, but is added to a new discussions app in openedx.core for now.

**Reviewers**
- [ ] @Agrendalath 
- [ ] @taniwha
- [ ] edX reviewer[s] TBD

**Settings**
```yaml
EDXAPP_FEATURES:
  ENABLE_COMBINED_LOGIN_REGISTRATION: true
```

edx-webhook commented:
Thanks for the pull request, @xitij2000! I've created [OSPR-4682](https://openedx.atlassian.net/browse/OSPR-4682) to keep track of it in JIRA, where we prioritize reviews.  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

Please let us know once your PR is ready for our review and all tests are green.

xitij2000 commented:
Certainly.

taniwha commented on openedx/core/djangoapps/config_model_utils/models.py:
```diff
@@ -316,30 +379,218 @@ def cache_key_name(cls, site, org, org_course, course_key):  # pylint: disable=a
-    @classmethod
-    def _org_from_org_course(cls, org_course):
-        return org_course.partition('+')[0]
+    def clean(self):
```
method name does not seem to reflect the code

xitij2000 replied:
@taniwha Could you expand on that a bit? I'm not sure I understand what the issue is.

taniwha commented:
#
:+1:

- [x] I read through the code
- [x] Includes documentation

It all looked pretty good to me. I took a look at the failing test, and while I'm not up to speed on the test system, it does seem odd that the first test ran without issues but the second failed despite being very similar.

Agrendalath commented on openedx/core/djangoapps/discussions/api/config.py:
```diff
@@ -0,0 +1,54 @@
+from ..models import DiscussionConfig, LearningContextDiscussionConfig
+
+
+def get_discussion_config_options(context_key: LearningContextKey) -> List[DiscussionConfigData]:
```
Should we introduce some tests for these two functions from this module?

Agrendalath commented on openedx/core/djangoapps/config_model_utils/tests/test_appconfig.py:
```diff
@@ -0,0 +1,176 @@
+
+    @ddt.data(
+        # site, org, org_course, context_key, slug_keys
+        (1, None, None, None, [1]),
+        (2, None, None, None, [2]),
```
Should we use `self.site1.id` and `self.site2.id`?

xitij2000 replied:
This is a decorator so exists before self is available. I will look into this. This test seemed to pass locally, not sure what the issue is here.

xitij2000 commented on lms/envs/common.py:
```diff
@@ -2379,6 +2379,7 @@ def _make_locale_paths(settings):  # pylint: disable=missing-function-docstring
 
     # Discussion forums
     'openedx.core.djangoapps.django_comment_common',
+    'openedx.core.djangoapps.discussions',
```
It should. Will do.

xitij2000 commented on openedx/core/djangoapps/discussions/__init__.py:
```diff
@@ -0,0 +1 @@
+default_app_config = 'openedx.core.djangoapps.discussions.apps.DiscussionsConfig'
```
That's a leftover from a previous version, will remove.

xitij2000 commented on openedx/core/djangoapps/config_model_utils/models.py:
```diff
@@ -316,30 +378,197 @@ def cache_key_name(cls, site, org, org_course, course_key):  # pylint: disable=a
+            MyCourseAppConfig.current(org_course='HarvardX/CS50')
+                # Includes global options, HarvardX's options and options for this specific course
+                [ <internal_forum>, <foruma>, <forumb>, <forumc> ]
+            MyCourseAppConfig.available(course_key=CourseKey(org='HarvardX', course='CS50', run='2018_Q1'))
+                # Includes global options, HarvardX's options and options for this specific course
+                [ <internal_forum>, <foruma>, <forumb>, <forumc> ]
+            MyCourseAppConfig.available(
+                course_key=CourseKey(org='HarvardX', course='CS50', run='2019_Q1')
+            )
+                # Includes global options, HarvardX's options and options for this specific course
+                [ <internal_forum>, <foruma>, <forumb>, <forumc> ]
+            bio101 = CourseKey(org='HarvardX', course='Bio101', run='2018_Q1')
+            MyCourseAppConfig.available(course_key=bio101)
```
Great catch! Earlier I was using coruse_key but made is more generic later. Will fix.

xitij2000 commented:
@stvstnfrd This is ready for review. I'm fully expecting the tests to pass, but will take a look in case they don't.

stvstnfrd commented on openedx/core/djangoapps/discussions/models.py:
```diff
@@ -0,0 +1,53 @@
+        default={},
+        help_text=_("The discussion configuration data that can be user visible."),
+    )
+    private_config = JSONField(
```
Can you provide some examples on how this will be used?

Will the private and public dictionaries be merged? Which takes precedence?
Will "child" configs (say, the full org, course, _and_ run are specified) "inherit" values from the "parent" (say, only org/site is specified) or will they just be used to "seed" the child at create-time (say, copy the settings from parent to child).

I see the general idea of what we're getting at w/ this hierarchy, but I think I'm still missing a few pieces...

xitij2000 replied:
I created two separate public and private configuration objects since I wanted to segregate secret configuration such as API keys, client keys/secrets and general configuration settings, such as project ID, etc. This allows a admin will access to secrets to set these up, and for other users to modify non-secret configuration without needing access to or seeing the secret config.

As such these will not be merged, and plugins will get them as separate values so they can handle them separately. 

There is no inheritance of values, each edit produces a new config and the final value is used. Eventually you will be able to clone a config and make a new copy with a new unique ID, but that will break the link. 

I think many plugins won't need the non-secret config since course-related config will be stored elsewhere, and I can understand eliminating it for now and only adding it in the future if a need arises. 

Example use:

    For Discourse the configurations could be as follows:
        Secret Config: { 
            api_key: 6f2343a1-d7d5-4994-a86b-a693fd980937
        }
        Config: {
            base_url: https://discourse.somesite.com
            auth_mode: LTI/SSO
            sync_moderators: true
            course_admins_can_moderate: false
        }

stvstnfrd replied:
Cool, in that case, as long as we have the option to extend in the future as needed, I think we can punt on the separate config for now, to avoid the complication (both code + UI/UX)

xitij2000 commented on openedx/core/djangoapps/config_model_utils/models.py:
```diff
@@ -45,11 +48,68 @@ def validate_course_in_org(value):
+        return NoneToEmptyQuerySet(self.model, using=self._db)
+
+
+def _org_from_org_course(org_course):
```
They are just extracted from static methods. I will check again to make sure I didn't change anything.

xitij2000 commented on openedx/core/djangoapps/discussions/api/config.py:
```diff
@@ -0,0 +1,54 @@
+        Returns `None` if a discussion tool isn't configures for the course yet, of if it is disabled.
+
+    """
+    # TODO: Check if discussion_enabled is set for the course. If not return None.
```
This is dependent on: https://github.com/edx/edx-platform/pull/24380

xitij2000 commented on openedx/core/djangoapps/config_model_utils/models.py:
```diff
@@ -234,7 +295,8 @@ def sort_key(override):
                         provenances[field.name] = Provenance.global_
 
         current = cls(**values)
-        current.provenances = {field.name: provenances[field.name] for field in stackable_fields}  # pylint: disable=attribute-defined-outside-init
```
It's just a line-wrap.

xitij2000 commented on openedx/core/djangoapps/config_model_utils/models.py:
```diff
@@ -167,13 +227,13 @@ def current(cls, site=None, org=None, org_course=None, course_key=None):  # pyli
 
         if org_course is None and course_key is not None:
-            org_course = cls._org_course_from_course_key(course_key)
+            org, org_course = _org_course_from_context_key(course_key)
```
I don't think a non-None value will ever be overwritten since:
- We check earlier in the method that only one of `site`, `org`, `org_course`, and `course_key` are provided. 
- So if `course_key` is not `None`, then `org` definitely is.

stvstnfrd replied:
Alright, I think I see it now; I guess the (pre-existing) logic is over-complicated (eg, above, that existing check could be `if  course_key is not None:`, since only one value can be non-None).

Nothing to see here!

xitij2000 commented on openedx/core/djangoapps/config_model_utils/models.py:
```diff
@@ -167,13 +227,13 @@ def current(cls, site=None, org=None, org_course=None, course_key=None):  # pyli
 
         if org_course is None and course_key is not None:
-            org_course = cls._org_course_from_course_key(course_key)
+            org, org_course = _org_course_from_context_key(course_key)
```
My assumption was that a CourseKey is a subset/type of context key so this would be OK, but I get your point from the previous comment, and will change accordingly.

xitij2000 commented on openedx/core/djangoapps/config_model_utils/models.py:
```diff
@@ -45,11 +48,68 @@ def validate_course_in_org(value):
+    return _org_course_from_context_key(course_key)[1]
+
+
+def _org_course_from_context_key(context_key):
```
Yes, currently I'm only aware of Coruses having an org, but keeping this for future cases.

xitij2000 commented:
While I did use some of the querying logic from `StackedConfiguration`, they are essentially very different and this is serving a completely different purpose. The aim of the stacked configuration is to allow for a base configuration that can then be overridden at different levels, i.e. you can have a config option set to `value1` globally, `value2` for a particular org, `value3` for a particular org_course etc.  In this case it will stack those configurations over each other such that the configurations are merged into the final state depending on which course, org_course, org, site etc they are accessed for. i.e. these different configs will be squashed into one based on where the config is accessed. 

Also one StackedConfiguraiton set will collapse down to a single configuration instance. In that case you will need a separate model for each plugin (which is probably OK) but then each plugin will only support one global config.

That is not what I'm going for here. I think a better way to describe this is via permissions. In this model a config object that only has a site set should be usable by anyone and any course in that site. A config object that has an org set should be usable by anyone and any course in that org, etc. So if you query the available configs for a particular course, it will return the configurations that are globally defined, defined for the course's site, it's or, its org_course etc. 

I think the "slug" is badly named here, "config_id" is probably a better name. Like all Configuration Models this too creates a new object on each modification, however the "slug" is retained across edits. So if an external model references a particular config "slug" it can fetch the latest version of that configuration. So if someone updates a configuration in one place, every place it's used will get the latest value.

xitij2000 commented:
jenkins run python

xitij2000 commented:
jenkins run py38 python

stvstnfrd commented on openedx/core/djangoapps/discussions/models.py:
```diff
@@ -0,0 +1,53 @@
+    )
+
+
+class LearningContextDiscussionConfig(models.Model):
```
TODO: @stvstnfrd to wrap head around use of this.

xitij2000 commented:
@stvstnfrd I've updated this PR, but am currently not rippling these changes to all the other PRs till this work is approved. 

Do tell if I need to make major changes to the approach here. We had discussed a few such as:
- Using [Django simple history](https://django-simple-history.readthedocs.io/en/latest/). This is currently used by the CourseOverview and some other newer models, but not by the standard configuration models. 
- Not bother with history at all. 

Both of these will require bigger changes, but will result in somewhat lesser code, and we can then use regular ids to link to a config object rather than indirectly using a config key.

xitij2000 commented:
@stvstnfrd I've added an ADR to this PR.

xitij2000 commented:
Based on today's discussion I will rewrite this to use django simple history instead.

bradenmacdonald commented on openedx/core/djangoapps/discussions/models.py:
```diff
@@ -0,0 +1,55 @@
+
+class LearningContextDiscussionConfig(models.Model):
+    """
+    Associates each learning context with a :class:`DiscussionConfiguration` via the config_key.
```
Nit: I recommend writing "each learning context **(e.g. each course run)** with a..." to make it clear that generally "learning context" = "course"

bradenmacdonald commented on openedx/core/djangoapps/discussions/models.py:
```diff
@@ -0,0 +1,55 @@
+    config = JSONField(
+        blank=False,
+        default={},
+        help_text=_("The discussion configuration data that can be user visible."),
```
Where does the data go that cannot be user visible? And does "user" refer to (some) admins, or any user at all?

bradenmacdonald commented on openedx/core/djangoapps/config_model_utils/models.py:
```diff
@@ -316,30 +379,215 @@ def cache_key_name(cls, site, org, org_course, course_key):  # pylint: disable=a
+class CourseAppConfigOptionsModel(ConfigurationModel):
+    """
+    A ConfigurationModel that stacks Global, Site, Org, Course, and Course Run level
+    configuration values.
```
Same docstring as `StackedConfigurationModel`

bradenmacdonald commented on openedx/core/djangoapps/config_model_utils/docs/decisions/0001-course-app-config-model.rst:
```diff
@@ -0,0 +1,85 @@
+The new model instead treats each configuration as a separate option available
+to course authors. So an org level configuration will not override a site level
+configuration, but will co-exist with it. Course authors will have both options
+at their disposal when configuring a course.
```
> an org level configuration will not override a site level configuration, but will co-exist with it. Course authors will have both options at their disposal when configuring a course.

Is this really necessary though?

You could get much of the same effect using `StackedConfigurationModel` if each plugin declared its own StackedConfigurationModel subclass, and each configuration option was a separate field (defined in `STACKABLE_FIELDS`); though it wouldn't allow a course to pick a site-level configuration of a given plugin instead of an org-level configuration, since the org-level values would always take precedence.

If any provider (like Discourse) is configured at the course/run level, other levels are irrelevant. Obviously I want to use the options configured at the course level, perhaps inheriting an API key from a higher level; certainly I don't need a choice between the course-level config and the org-level config for a given provider, because the course-level config wouldn't exist if it wasn't the one I wanted to use in the course.

So really the only question is: is there ever a time when a provider (say Discourse) is configured at the Site level _and_ the Org level, but a particular course wants to use the site-level version, instead of the Org level version? Maybe, but that seems like a real edge case to me. (More realistic is an org migrating to their own config, and needing old courses to continue to inherit the site-level config)

---------------------------------------

Which brings me to another point: what happens when the site-level or org-level configuration is changed - won't it possibly interrupt many running or archived courses (any that don't specify a course-level config), possibly pointing their discussion board to a new service or different forum on the same service?

---------------------------------------

Does this include any UI / permissions to allow global admins to set global values for the instance, "site admins" to set values for the site, "org admins" to set values for the org, and course instructors to set values for the course?

If not, I'd suggest a much simpler implementation, using regular Django models:

```python
class DiscussionPluginConfiguration:
    """
    Configure a discussion plugin for this Open edX instance, optionally restricted to a site or org.

    Only global staff (sysadmins) can configure these. 

    The "config" can be blank, which means that the plugin is enabled for this Open edX installation/site/org. Multiple configurations may exist for the same plugin, and each course may choose which one to use.
    """
    name = String()  # Name of this configuration e.g. "SomeOrg Discourse"
    plugin = String()  # Which implementation: cs_comments_service, Discourse,  etc.
    config = JsonField()  # API key, endpoint, etc. Nobody other than global staff can view the config data, as it may contain secrets like API keys.
    restrict_to_site = ForeignKey  # Only courses associated with this site(s) may use this config
    restrict_to_org = ForeignKey  # Only courses associated with this organization(s) may use this config

class CourseDiscussionConfiguration:
    """
    Configure a discussion plugin used by a specific course / learning context.
    If an entry is not found for a given course/context, discussions are disabled.
    """
    context_key = LearningContextKey()  # Which course/context this discussion config is for
    enabled = BooleanField()
    global_config = ForeignKey(DiscussionPluginConfiguration)
    course_config = JsonField()  # Overrides like course-specific settings, API key (if different, or if not specified globally), etc.
```

This approach:
* Requires no new abstractions
* Makes minimal use of inheritance
  - site/org is only checked as a permissions/security/validation check when an update to a given course's `CourseDiscussionConfiguration` is made. Inheritance checks are not needed when reading config.
  - This means no complex caching/inheritance logic is needed. The merged configuration for a given course can be queried from a single trivial `CourseDiscussionConfiguration.objects.select_related('global_config').get(context_key=context_key)`
* Has a very clear permissions scheme: only global admins can enable a plugin system-wide or for a particular course/org; they can also (optionally) set default secrets like API key, OAuth creds, etc. but don't have to. Only global admins can see the config values in the `DiscussionPluginConfiguration`, and only course staff can see the values in the `CourseDiscussionConfiguration`.
* Each course always explicitly selects a plugin+configuration to use. If the system-wide configuration for the plugin already sets the required API key etc., then no additional configuration is needed, but the course can always set their own API key or override any other options as needed.
   - This is important to ensure that actions like adding a new org-level configuration don't suddenly inherit down and change what forum is being used for a running course.

I'm not married to this approach, just wanted to suggest it as this PR seems very complex. In any case I'd also suggest starting with a simple configuration paradigm and getting the actual discussion plugins and UI working and implemented before spending a lot more time on configuration helpers and inheritance. The configuration needs may change once more stuff is implemented.

xitij2000 replied:
> You could get much of the same effect using `StackedConfigurationModel` if each plugin/provider (Discourse, etc.) declared its own StackedConfigurationModel subclass, and each configuration option was a separate field (defined in `STACKABLE_FIELDS`); though it wouldn't allow a course to pick a site-level configuration of a given plugin instead of an org-level configuration, since the org-level values would always take precedence.

StackedConfiguration does too much. It automatically stacks and merges fields across multiples levels and that is not what we want here. Simply taking the highest precedence version is enough. 

> Which brings me to another point: what happens when the site-level or org-level configuration is changed - won't it possibly interrupt many running or archived courses (any that don't specify a course-level config), possibly pointing their discussion board to a new service or different forum on the same service?

This is by design, to allow a changing a config (such as the API key for a service) in one place and have it apply across all services. The plugin that a config is linked to can't be changed, however the URL might. In this case though, that might be desired. If it isn't desired, then we need to warn when performing such an action ("You are modifying a configuration used by X active courses, are you sure you want to continue?"). 

> Does this include any UI / permissions to allow global admins to set global values for the instance, "site admins" to set values for the site, "org admins" to set values for the org, and course instructors to set values for the course?

Not yet. Currently, superadmins with access to this page can set it for all levels. The rest can come as a future enhancement. 

> Each course always explicitly selects a plugin+configuration to use. If the system-wide configuration for the plugin already sets the required API key etc., then no additional configuration is needed, but the course can always set their own API key or override any other options as needed. 

This is already true of the current PR, I've called it a provider instead of plugin, but that sets which plugin this config is tied to, and it cannot be changed. 

The problem with any kind of merging (and the reason why I didn't just use stackedconfig) is that only the plugin can know how the configuration is structured, and can be merged. Perhaps the plugin can provide its own merge function that makes it complicated for an admin to understand how the `course_config` override will work, or the `course_config` can just override entirely (i.e. mask) the `global_config` in which case it might make more sense to make these an either/or option, either pick a pre-existing config, or use a custom config, in which case how do you specify a plugin?

bradenmacdonald replied:
Ok great.

My main suggestion was not to use StackedConfiguration, but the two-level approach I suggested, which doesn't require [a new abstraction](https://github.com/edx/edx-platform/blob/24676099d84f5f94e0bcf2d3e1bac0081ba71dd9/openedx/core/djangoapps/config_model_utils/models.py#L390-L394) nor [complex database queries](https://github.com/edx/edx-platform/blob/24676099d84f5f94e0bcf2d3e1bac0081ba71dd9/openedx/core/djangoapps/config_model_utils/models.py#L569-L591), and which allows out-of-the box use of the django admin UI by separating admin-level settings and user-level settings.

> The problem with any kind of merging (and the reason why I didn't just use stackedconfig) is that only the plugin can know how the configuration is structured, and can be merged.

Re merging, I think if you just say that any top-level keys defined in the `CourseDiscussionConfiguration.course_config` will completely overwrite any similar keys in the higher-level `DiscussionPluginConfiguration.config`, if any, then that is workable with any plugin and plugin authors can structure their config format accordingly, and users/admins will understand because it's very clear and works exactly the same way as other configuration keys in the system. (Like `private.py` which overrides `lms.yml` which overrides `common.py`, and in each case the top level keys completely replaces those of previous levels. This is exactly how all django settings work in general, how other [django app plugins](https://github.com/edx/edx-django-utils/tree/master/edx_django_utils/plugins) settings work, and so on.)

Besides, if users override a field like "API key" and forget to also override "endpoint URL", that's a simple error they can fix - it's not something we need to enforce in the configuration system design, by requiring those both to be set at the same time or neither.

However, if what you've got works, and you think the cost of introducing a new abstraction is offset by the benefits and future additional uses, then go with that.

xitij2000 replied:
@bradenmacdonald I updated this PR to use your suggested approach last sprint.

bradenmacdonald replied:
Cool, hope it works well 🤞

xitij2000 commented:
@stvstnfrd  I've rewritten and simplified the approach significantly based on @bradenmacdonald's approach above. I've also updated the ADR accordingly. 

Once this is reviewed, I will rebase some of the other dependent PRs onto this. Since this PR exports an API and I mainly use that API I don't expect too many ripple effects even though the approach has changes significantly.

davidjoy commented on openedx/core/djangoapps/discussions/docs/decisions/0001-discussion-configuration-api.rst:
```diff
@@ -0,0 +1,65 @@
+be the id of the plugin for which the configuration is defined), and a config
+JSON field (this will store the actual config). A configuration can be also be
+restricted to only be visible/available for a specific site, or org by setting
+`restrict_to_site` or `restrict_to_org`. Only one of these can be set at a time
```
Thinking ahead a bit - what do you think the effort would be to add other types of restrictions in the future, given the current model?  i.e., `restrict_to_program`, for instance.  I'm not at all saying we need to do this now, just thinking about future extensibility as we have shared discussions forums for various types of groups.

xitij2000 replied:
I think that will depend on how such future content is structured. My understanding is more of these relationships will be formalised in the database going forwards, so adding more restrictions should be straightforward. I only see complexity if we want to allow some combination of restrictions, but not others. 

For the specific case of programs since that relationship is in an external DB it definitely will complicate these queries. If the end goal is to have the same discussion provider and discussion instance across all courses in the program, then that would be accomplished by linking a `LearningContextDiscussionConfig` to the entire program via a program key. 

If the goal instead is that the config should only be available for use for courses inside a program, then perhaps it might make sense to cache the program key course key relation in edx-platform as well.

davidjoy commented on openedx/core/djangoapps/discussions/docs/decisions/0001-discussion-configuration-api.rst:
```diff
@@ -0,0 +1,65 @@
+The `DiscussionProviderConfig` model stores a configuration name (this will be
+visible to course authors configuring a discussion tool), a provider (this will
+be the id of the plugin for which the configuration is defined), and a config
+JSON field (this will store the actual config). A configuration can be also be
```
(This may be/probably should be beyond the scope of this PR... just something I was thinking about as I read)

xitij2000 replied:
Currently, there will only be the django admin site that can be edited as a raw text field. It will be validated as JSON but not whether it matches the scheme expected by a discussion app. 

That said, once we have a plugin API, the validation can be offloaded to the plugin via a standard API, if the plugin supports it. i.e. the plugin can expose a `validate_config` function that if present will be used to check if the configuration is valid.

There will also be a UI to edit these settings in the course-authoring MFE, however that will only allow editing the configuration specific to a course, not the configuration available globally. That said, once the UI already exists, it could be reused to create an admin UI in course-authoring that allows editing these base configuration sets in the future.

davidjoy replied:
This all makes sense to me, thanks!

davidjoy commented on openedx/core/djangoapps/discussions/migrations/0001_initial.py:
```diff
@@ -0,0 +1,101 @@
+                ('name', models.CharField(help_text='A user-friendly name for this configuration. e.g. SomeOrg Discourse', max_length=100)),
+                ('provider', models.CharField(db_index=True, help_text='The discussion tool/provider.', max_length=100, verbose_name='Discussion provider')),
+                ('config', jsonfield.fields.JSONField(default={}, dump_kwargs={'cls': jsonfield.encoder.JSONEncoder, 'separators': (',', ':')}, help_text='The configuration data for this provider.', load_kwargs={})),
+                ('restrict_to_org', models.ForeignKey(blank=True, help_text='Optionally restrict this config for use only on this Organization.', null=True, on_delete=django.db.models.deletion.CASCADE, to='organizations.Organization')),
```
Just to sanity check this - we're saying that when an org or site is deleted, we should delete all discussion provider configs that referenced it.

I _think_ that makes sense... is there any situation where we would want one of these configs to survive an associated org or site?  Should the foreign key be set to null so that the config persists and goes back into the general population of configs, minus its restriction?

xitij2000 replied:
I thought about this, but given the way it works, I didn't feel comfortable not cascading the deletion. If we set it to null, that will remove the restriction as you mentioned, but that may mean that a configuration (which contains secret credentials) will now become globally available. 

Even though only admins will be able to see it either way, it will become visible as an option to other course staff, which may not be desirable. Other options I considered were: add a setting to switch over to a new default org on delete, and add a signal to disable the config if the org is deleted. 

Finally though, I realised that since we're already using django simple history here, it will maintain historical records of the model even if it's deleted, so at least the data isn't lost. If a situation comes up where a delete needs to be reverted the historical record can do that.

davidjoy replied:
I think this makes sense - thanks for talking it through!

davidjoy commented on openedx/core/djangoapps/discussions/docs/decisions/0001-discussion-configuration-api.rst:
```diff
@@ -0,0 +1,65 @@
+an option. Changes to this model are tracked using Django Simple History.
+
+To actually use a configuration for a particular course, we use the
+`LearningContextDiscussionConfig` model. This will link a course to a particular
```
Should this be specifying "course" here/is that what it's actually associated with?  Will we have different types of configs for associating a provider with a different type of entity?  I.e., configs for sites, programs, orgs, sequences, or units, for instance?

xitij2000 replied:
I'm using learning context keys here mainly for future-proofing, but currently courses are the only supported context, and as such they are the ones I refer to here. 

If in the future I definitely see a case for associating it with other kinds of learning contexts, but this mechanism will likely not be used for sites. If programs are represented as a learning context, then they can be  linked here. If one learning context can contain another, then we'll need a some more complex mechanism to resolve the config, that first checks the immediate learning context (the course) and then its parent (the program) etc.

davidjoy commented on openedx/core/djangoapps/discussions/models.py:
```diff
@@ -0,0 +1,113 @@
+        help_text=_("The configuration to use for this learning context."),
+        on_delete=models.SET_NULL,
+    )
+    config_overrides = JSONField(
```
Can we talk a little about these `config_overrides` and how they relate to `config`?

i.e., can we give some examples of what will likely be in `config`, and what likely goes in `config_overrides`?  

I'm trying to understand how common it'll be to override config at the learning context level, and whether we're likely to override just _part_ of it, whether most/all of it would commonly be overridden, or whether it's possible to make a non-sensical configuration by overriding only a part of a set of fields, such as overriding a URL/tool identifier but not a secret key, for instance, making a configuration with the wrong key for the service.

Do we have known use cases for needing the overrides?  Or are we building in flexibility in "just in case"?  (I'm assuming the former, but don't have insight into the particular fields that would go into these configs...)

davidjoy replied:
To summarize the above:

1. I'd love to understand the detailed fields we expect to go into `config`
2. And what we expect might go into `config_overrides`.
3. And the common usage patterns between the two.

xitij2000 replied:
That will depend on the specific plugin, but looking at Piazza, and YelloDig which both integrate via LTI. It could be:
- Consumer Key
- Consumer Secret

For Discourse it might be:
- API key
- Discourse instance url
- Enable attachments
- embed mode: lti / iframe etc

In such a case a global config could provide all the above options, and a config override could choose to override it on a course-level basis. It might not make sense for a course to override the API configuration or the discourse instance (those changes should result in a new base configuration) but there could be other settings that could be changed. 

It will be the responsibility of the plugin to disallow any changes on a course level that don't make sense. The LTI credentials for instance won't be visible to course authors at all. In that sense `course_overrides` can probably be considered more like `course_config` i.e. config options that are course-specific, however it will still shadow the global config (so that it can provide defaults) so calling it overrides is still accurate. 

I don't expect course authors to ever know that they are overriding anything that all will be handled transparently for them. Even the API that plugins ~will~ should use will not expose these separately (at least that is the intention in this PR). A plugin can request or update the course-level config and it will just know the config for that particular course without being aware of the base/parent config it's based on. This allows for some refactoring of how exactly this works behinds the scenes without breaking API.

xitij2000 commented:
@stvstnfrd @davidjoy I'd love to know how I can get this merge. Are there any concerns with the current implementation that I have to address?

xitij2000 commented:
jenkins run js

stvstnfrd commented:
@xitij2000 We can close this in favor of https://github.com/edx/edx-platform/pull/25837, right?

xitij2000 commented:
> @xitij2000 We can close this in favor of #25837, right?

Yes.
